### PR TITLE
Issue #10134: Incorrect catch parameter line/ column number reported

### LINKED ISSUE #10134: Incorrect catch parameter line/ column number reported
Checkstyle reports that the catch parameters for a multicatch ast begin at the position of the `BOR`, because we make this the root of the multicatch tree:

```
➜  src javac Test.java
Note: Test.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

➜  src cat Test.java 
import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public class Test {
    public static void main(String[] args) {
        try {
            System.out.println(args[7]);
            File myFile = new File("myfile.txt");
            InputStream stream = myFile.toURL().openStream();
        } catch (ArrayIndexOutOfBoundsException | IOException e) {

        }
    }

}

➜  src java -jar checkstyle-8.43-all.jar -t Test.java
...
`--LITERAL_CATCH -> catch [11:10]
    |--LPAREN -> ( [11:16]
    |--PARAMETER_DEF -> PARAMETER_DEF [11:48]
    |   |--MODIFIERS -> MODIFIERS [11:48]
    |   |--TYPE -> TYPE [11:48]
    |   |   `--BOR -> | [11:48]
    |   |       |--IDENT -> ArrayIndexOutOfBoundsException [11:17]
    |   |       `--IDENT -> IOException [11:50]
    |   `--IDENT -> e [11:62]
    |--RPAREN -> ) [11:63]
    `--SLIST -> { [11:65]
    |       |           `--RCURLY -> } [13:8]
...

```

We should allow the qualified names and `|` to be parallel elements in the AST, which corresponds with Java grammar:

```
CatchType:
    QualifiedIdentifier { | QualifiedIdentifier }

```

Doing so allows us to report the correct line and column numbers in the multicatch parameter AST (and traverse the qualified names by iteration, instead of recursion):

```
`--LITERAL_CATCH -> catch [11:10]
    |--LPAREN -> ( [11:16]
    |--PARAMETER_DEF -> PARAMETER_DEF [11:17]
    |   |--MODIFIERS -> MODIFIERS [11:17]
    |   |--TYPE -> TYPE [11:17]
    |   |   |--IDENT -> ArrayIndexOutOfBoundsException [11:17]
    |   |   |--BOR -> | [11:48]
    |   |   `--IDENT -> IOException [11:50]
    |   `--IDENT -> e [11:62]
    |--RPAREN -> ) [11:63]
    `--SLIST -> { [11:65]
        `--RCURLY -> } [13:8]

```

-------

Ideally AST should looks same as `a || b || c` in logical expressions [link](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LOR)
```
  |--EXPR -> EXPR
  |   `--LOR -> ||
  |       |--IDENT -> a
  |       `--IDENT -> b
```
But in logical expression priority of apply matters for  example `a && b || c`, it is a bit critical for logical expressions to be like this.

list of types in catch is more like https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#TYPE_PARAMETERS
```
 |--TYPE_PARAMETERS -> TYPE_PARAMETERS
 |   |--GENERIC_START -> <
 |   |--TYPE_PARAMETER -> TYPE_PARAMETER
 |   |   `--IDENT -> A
 |   |--COMMA -> ,
 |   |--TYPE_PARAMETER -> TYPE_PARAMETER
 |   |   `--IDENT -> B
``` 

where separator is NOT a parent but sibling to elements.
Catch types are not bound by operator, they are list of types, so it is ok to make AST structure sequential in this case.

### PR DESCRIPTION
**Pull request by nmancus1**
Closes #10134 

### Check Regression Reports

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_checks-nonjavadoc-error/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_checks-only-javadoc-error/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part1/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part2/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part3/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part4/index.html

Fixed a false positive in `ClassFanOutComplexity`:

Example from https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part4/guava/xref/home/nick/development/contribution/checkstyle-tester/repositories/guava/guava/src/com/google/common/collect/Sets.java.html#L69

From master:

```
Referenced classes:
[AtomicHelper, Beta, CanIgnoreReturnValue, Cancellation, CancellationException, ExecutionException, Executor, Failure, ForOverride, Future, GwtCompatible, InternalFutureFailureAccess, InterruptedException, ListenableFuture, Listener, Logger, NullableDecl, ReflectionSupport, Runnable, SafeAtomicHelper, ScheduledFuture, SetFuture, SynchronizedHelper, TimeUnit, TimeoutException, Trusted, UnsafeAtomicHelper, V, Waiter, |]

```
Note the last "class". It is a `|`.

From `issue-10134`:

```
Referenced classes:
[AtomicHelper, Beta, CanIgnoreReturnValue, Cancellation, CancellationException, ExecutionException, Executor, Failure, ForOverride, Future, GwtCompatible, InternalFutureFailureAccess, InterruptedException, ListenableFuture, Listener, Logger, NullableDecl, ReflectionSupport, Runnable, SafeAtomicHelper, ScheduledFuture, SetFuture, SynchronizedHelper, TimeUnit, TimeoutException, Trusted, UnsafeAtomicHelper, V, Waiter]

```
The `|` is gone. 

It seems that `AbstractClassCouplingCheck` does not handle multicatch parameters properly, i.e. does not count any referenced class after the first one. Issue opened at https://github.com/checkstyle/checkstyle/issues/10144

Unit test added at https://github.com/checkstyle/checkstyle/blob/97bb4fbbcd187ada7f95f5e974759cdac2bbfa1e/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part5/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_part6/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_sevntu-check-regression_part_1/index.html

Clean

----------------------------

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/diff_sevntu-check-regression_part_2/index.html

New violations found for `IllegalCatchExtended`.

----------------------------

### ANTLR Regression Report

https://nmancus1.github.io/issue-10134_check_diff_reports_2021_06_19/antlr-diff/index.html

All changes expected.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java [hunk 1]
```diff
@@ -212,16 +212,9 @@ public void visitToken(DetailAST detailAST) {
     private static List<DetailAST> getAllExceptionTypes(DetailAST parentToken) {
         DetailAST currentNode = parentToken.getFirstChild();
         final List<DetailAST> exceptionTypes = new LinkedList<>();
-        if (currentNode.getType() == TokenTypes.BOR) {
-            exceptionTypes.addAll(getAllExceptionTypes(currentNode));
-            currentNode = currentNode.getNextSibling();
+        while (currentNode != null) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java:215]: Please remove `BOR` from the Javadoc of this method.

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java:215]: Or, even better, remove this method. It was a workaround for the bug you are fixing and does nothing for now.
The method `visitToken` should use `getFirstChild`/`getNextSibling` directly.

**Review by pbludov — CHANGES_REQUESTED** [review]
One minor change please

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java [hunk 2]
```diff
@@ -87,11 +87,21 @@ public void testIllegalClassNamesBad() throws Exception {
     @Test
     public void testMultipleTypes() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(IllegalCatchCheck.class);
+        checkConfig.addAttribute("illegalClassNames",
+                                 "java.lang.Error, java.lang.Exception, NullPointerException,"
+                                + "OneMoreException, RuntimeException, SQLException");
```
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java:92]: @pbludov

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java:215]: Added some more `illegalClassNames` to show that we use this iterative approach to getting Exception class names: https://github.com/checkstyle/checkstyle/pull/10136/files#r659173444

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java [hunk 3]
```diff
@@ -87,11 +87,21 @@ public void testIllegalClassNamesBad() throws Exception {
...
             "7:11: " + getCheckMessage(MSG_KEY, "RuntimeException"),
+            "7:11: " + getCheckMessage(MSG_KEY, "SQLException"),
             "10:11: " + getCheckMessage(MSG_KEY, "RuntimeException"),
+            "10:11: " + getCheckMessage(MSG_KEY, "SQLException"),
+            "10:11: " + getCheckMessage(MSG_KEY, "OneMoreException"),
+            "13:11: " + getCheckMessage(MSG_KEY, "OneMoreException"),
             "13:11: " + getCheckMessage(MSG_KEY, "RuntimeException"),
+            "13:11: " + getCheckMessage(MSG_KEY, "SQLException"),
+            "16:11: " + getCheckMessage(MSG_KEY, "OneMoreException"),
+            "16:11: " + getCheckMessage(MSG_KEY, "SQLException"),
```
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java:104]: @pbludov

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java (hunk 1 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java:215]: I'm ok to keep this method, I just want to note that it copies from one linked list to another.

**Review by pbludov — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java]: I totally misunderstood what you meant; I have removed this method.  Thanks for pointing this out :)

**Review by pbludov — APPROVED** [review]

**Review by strkkk — APPROVED** [review]

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
Can we keep BOR to be parent for IDENT ?

In all other cases BOR rendering different AST.

From human perspective new ast is better, but for interpretation of code placing operator as parent is more useful. 
What is ast for multiple BOR ?

**Comment by pbludov** [comment]
> Can we keep BOR to be parent for IDENT ?
> In all other cases BOR rendering different AST.

This is because `BOR` is not a binary operator here. It is a separator for the list of types.
For an arithmetic expression like `a + b | c * d` thrre should be  a binary tree in the AST to iterate the expression in the right way.
For a list of exception types it make no sense, it just produce strange one-side tree:
`catch (First | Second | Third ex)` becomes
```
LITERAL_CATCH -> catch [9:6]
 |--LPAREN -> ( [9:12]
 |--PARAMETER_DEF -> PARAMETER_DEF [9:28]
 |   |--MODIFIERS -> MODIFIERS [9:28]
 |   |--TYPE -> TYPE [9:28]
 |   |   `--BOR -> | [9:28]
 |   |       |--BOR -> | [9:19]
 |   |       |   |--IDENT -> First [9:13]
 |   |       |   `--IDENT -> Second [9:21]
 |   |       `--IDENT -> Third [9:30]
 |   `--IDENT -> ex [9:36]
 |--RPAREN -> ) [9:38]
```
It looks as if there is some kind of hierarchy, as in the package paths for example, but there is nothing like that. This pseudo-hierarchy just makes things more complicated. 
This PT consists of 90% of removing strange and non-obvious code, like [this](https://github.com/checkstyle/checkstyle/pull/10136/files#diff-912a244f4240b3217ce9237404bd58f6fce41287a18efaba996a6937787c3e28L202).
By merging this PR, we will simplify the writing of third-party checks, making the processing of `BOR` unnecessary in each of the checks for the `catch` entry.

**Comment by nmancus1** [comment]
Adding to @pbludov's comment; from https://docs.oracle.com/javase/specs/jls/se16/html/jls-14.html#jls-Catches: 

> An exception parameter may denote its type as either a single class type or a union of two or more class types (called alternatives). The alternatives of a union are syntactically separated by |.

Treating the `BOR` as an operator in this case would raise questions like: is `BOR` in this case right or left associative? 

> Previously we look like user ast that is interpretation focused

I think that creating a "list of alternatives" rather than a "tree of alternatives" for the multicatch statement is an accurate interpretation of the JLS.

**Comment by romani** [comment]
@nmancus1 , please rebase to resolve conflict.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/* Config:
+ * max = "2"
+ */
+public class InputClassFanOutComplexityRemoveMultiCatchBitwiseOr { // ok
+    public static void main(String[] args) {
+        try {
+            System.out.println(args[7]);
+            File myFile = new File("myfile.txt");
+            InputStream stream = myFile.toURL().openStream();
+        } catch (ArrayIndexOutOfBoundsException | IOException e) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java:16]: please make catch with 3 of `|`.

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputRegressionJavaClass1Ast.txt [hunk 5]
```diff
@@ -2219,12 +2219,12 @@ CLASS_DEF -> CLASS_DEF [17:0]
...
-    |       |   |   |   |--TYPE -> TYPE [312:29]
-    |       |   |   |   |   `--BOR -> | [312:29]
-    |       |   |   |   |       |--IDENT -> IOException [312:17]
-    |       |   |   |   |       `--IDENT -> ArrayIndexOutOfBoundsException [312:31]
+    |       |   |   |--PARAMETER_DEF -> PARAMETER_DEF [312:17]
+    |       |   |   |   |--MODIFIERS -> MODIFIERS [312:17]
+    |       |   |   |   |--TYPE -> TYPE [312:17]
+    |       |   |   |   |   |--IDENT -> IOException [312:17]
+    |       |   |   |   |   |--BOR -> | [312:29]
+    |       |   |   |   |   `--IDENT -> ArrayIndexOutOfBoundsException [312:31]
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputRegressionJavaClass1Ast.txt:2227]: please make catch with 3 of `|`.

Ideally AST should looks same as `a || b || c` in logical expressions [link](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LOR)
```
  |--EXPR -> EXPR
  |   `--LOR -> ||
  |       |--IDENT -> a
  |       `--IDENT -> b
```
But in logical expression priority of apply matters for  example `a && b || c`, it is a bit critical for logical expressions to be like this.

list of types in catch is more like https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#TYPE_PARAMETERS
```
 |--TYPE_PARAMETERS -> TYPE_PARAMETERS
 |   |--GENERIC_START -> <
 |   |--TYPE_PARAMETER -> TYPE_PARAMETER
 |   |   `--IDENT -> A
 |   |--COMMA -> ,
 |   |--TYPE_PARAMETER -> TYPE_PARAMETER
 |   |   `--IDENT -> B
``` 

where separator is NOT a parent but sibling to elements.
Catch types are not bound by operator, they are list of types, so it is ok to make AST structure as you suggested.

ATTENTION: if you agree, please copy this explanation to issue description to exlpain rationale to users.
This issue is breaking compatibility.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java (hunk 4 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java]: Added 4 for good measure.

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputAstRegressionManyAlternativesInMultiCatch.txt [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    |       |       |   |   |--IDENT -> ArrayIndexOutOfBoundsException [14:17]
+    |       |       |   |   |--BOR -> | [14:48]
+    |       |       |   |   |--IDENT -> IOException [14:50]
+    |       |       |   |   |--BOR -> | [15:16]
+    |       |       |   |   |--IDENT -> SQLException [15:18]
+    |       |       |   |   |--BOR -> | [15:31]
+    |       |       |   |   |--IDENT -> SecurityException [15:33]
+    |       |       |   |   |--BOR -> | [15:51]
+    |       |       |   |   `--IDENT -> OneMoreException [15:53]
+    |       |       |   `--IDENT -> e [15:70]
```
Comment (nmancus1) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputAstRegressionManyAlternativesInMultiCatch.txt:132]: @romani longer list of alternatives here

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputRegressionJavaClass1Ast.txt (hunk 5 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputRegressionJavaClass1Ast.txt:2226]: > please make catch with 3 of `|`.

https://github.com/checkstyle/checkstyle/pull/10136/files#r671704942

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputRegressionJavaClass1Ast.txt:2226]: I updated description

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java [hunk 7]
```diff
@@ -87,11 +87,21 @@ public void testIllegalClassNamesBad() throws Exception {
     @Test
     public void testMultipleTypes() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(IllegalCatchCheck.class);
+        checkConfig.addAttribute("illegalClassNames",
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java:90]: please use `addProperty`, we just merged update for this class, and Configuration.xxxAttributeXxxx are deprecated.

in all places

**Review by romani — CHANGES_REQUESTED** [review]
last minor:

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java (hunk 7 above)
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheckTest.java]: done, separate commit at https://github.com/checkstyle/checkstyle/pull/10136/commits/eb7af217b27a6dade42cd8eb955335f0cdd5cd7c to update usage from https://github.com/checkstyle/checkstyle/pull/10122

**Review by romani — APPROVED** [review]

**Pull request merged**